src/auth/permissions/index.ts

```diff
@@ -0,0 +1,173 @@
+import * as knex from 'knex';
+import { PermissionsInitialiser, GetPermissionIds, GetRoleId } from './permissions';
+
+const getPermissionIds: GetPermissionIds = ({ client, resource, permissionLevel, access }) =>
+  client
+    .select('permission_id')
+    .table('permission')
+    .where({
+      ['permission_entity']: resource,
+      ['permission_level']: permissionLevel,
+      ['access_type']: access,
+    });
+
+const getRoleId: GetRoleId = async({ client, role }) =>
+  client
+    .select('access_role_id')
+    .table('access_role')
+    .where({ ['access_role_name']: role })
+    .then((res) => res[0].access_role_id);
+
+const permissionsInitialiser: PermissionsInitialiser = (config) => {
+  const client = knex(config);
+  return {
+    grantExisting: async ({ resource, permissionLevel, access, role }) => {
+      const roleId = await getRoleId({ client, role });
+
+      const permissionIds = await getPermissionIds({
+        client,
+        resource,
+        permissionLevel,
+        access });
+
+      if (permissionIds.length === 1) {
```

Comment:
I think this line suggests we're missing something in our DB constraints. Shouldn't the combination of (resource, access, permission level) have a unique constraint? Would there be any case where we'd need to have two rows with the same 3-tuple?

Reply:
This is checking that the permission (resource x permissionLevel x access) exists. I'm missing an else statement that throws an error if the combination does not exist.

Reply:
Oh, OK. In which case the sub-query style that I mention in one of my later comments means this is probably unnecessary, since `knex` will throw if a sub-query returns `N !== 1` rows.

Reply:
I'll look into what the error messages from knex are like under, although I was going for module specific error messages which tell you why is went wrong and what you should use instead